packages/angular_devkit/build_angular/test/browser/web-worker_spec_large.ts

```diff
@@ -0,0 +1,179 @@
+        expect(workerContent).toContain('bar');
+
+        const mainContent = virtualFs.fileBufferToString(
+          host.scopedSync().read(join(outputPath, 'main.js')),
+        );
+        // main bundle references worker.
+        expect(mainContent).toContain('0.worker.js');
+      }),
+      // Doesn't show any warnings.
+      tap(() => expect(logger.includes('WARNING')).toBe(false, 'Should show no warnings.')),
+    ).toPromise().then(done, done.fail);
+  });
+
+  it('minimizes and hashes worker', (done) => {
+    host.writeMultipleFiles(workerFiles);
+    const overrides = {
+      webWorkerTsConfig: 'src/tsconfig.worker.json',
+      outputHashing: 'all',
+      optimization: true,
+    };
+    runTargetSpec(host, browserTargetSpec, overrides).pipe(
+      tap((buildEvent) => expect(buildEvent.success).toBe(true)),
+      tap(() => {
+        const workerBundle = host.fileMatchExists(outputPath,
+          /0\.[0-9a-f]{20}\.worker\.js/) as string;
+        expect(workerBundle).toBeTruthy('workerBundle should exist');
+        const workerContent = virtualFs.fileBufferToString(
+          host.scopedSync().read(join(outputPath, workerBundle)),
+        );
+        expect(workerContent).toContain('hello from worker');
+        expect(workerContent).toContain('bar');
+        expect(workerContent).toContain('"hello"===t&&postMessage(o.foo)');
+
+        const mainBundle = host.fileMatchExists(outputPath, /main\.[0-9a-f]{20}\.js/) as string;
+        expect(mainBundle).toBeTruthy('mainBundle should exist');
+        const mainContent = virtualFs.fileBufferToString(
+          host.scopedSync().read(join(outputPath, mainBundle)),
+        );
+        expect(mainContent).toContain(workerBundle);
+      }),
+    ).toPromise().then(done, done.fail);
+  });
+
+  it('rebuilds TS worker', (done) => {
+    host.writeMultipleFiles(workerFiles);
+    const overrides = {
+      webWorkerTsConfig: 'src/tsconfig.worker.json',
+      watch: true,
+    };
+
+    let buildCount = 0;
+    let phase = 1;
+    const workerPath = join(outputPath, '0.worker.js');
+    let workerContent = '';
+
+    runTargetSpec(host, browserTargetSpec, overrides, DefaultTimeout * 3).pipe(
+      // Wait for files to be written to disk.
+      debounceTime(1000),
+      tap((buildEvent) => expect(buildEvent.success).toBe(true, 'build should succeed')),
+      tap(() => {
+        buildCount++;
+        switch (phase) {
+          case 1:
+            // Original worker content should be there.
+            workerContent = virtualFs.fileBufferToString(host.scopedSync().read(workerPath));
+            expect(workerContent).toContain('bar');
+            // Change content of worker dependency.
+            host.writeMultipleFiles({ 'src/app/dep.ts': `export const foo = 'baz';` });
+            phase = 2;
+            break;
+
+          case 2:
+            // TODO(filipesilva): Should this change? Check with Jason Miller.
```

Comment:
@developit figured out the rebuild issue. It's because the worker changes name on rebuilds. Is this intended?

Reply:
I'm also seeing a weird behaviour where sometimes the child compilation picks up on changed files, and sometimes it doesn't.

In this test I have seen all these outputs just by running the rebuild test repeatedly with no code changes:
- webpack loader is called for `src/app/dep.ts` twice and registers the change from `bar` to `baz`, `baz` ends up in `0.worker.js`
- same as above but `baz` ends up in `1.worker.js`
- loader is only called once (initially), neither `0.worker.js` nor `1.worker.js` contain `baz`

Do you have any idea of what's happening here?

Reply:
I think the worker changing names may be somewhat unavoidable. There's a `filename` option that would make it possible to use a static name, but that doesn't seem like the right solution here either.

There's [no caching enabled](https://github.com/GoogleChromeLabs/worker-plugin/blob/master/src/loader.js#L27) in worker-plugin, so it's likely something deeper in Webpack.

I'm going to add some rebuild tests to the plugin to see if I can trigger this in isolation. My hunch is that the [parser hook's instance counter](https://github.com/GoogleChromeLabs/worker-plugin/blob/7ea61fb83a02170d0235e4590c4deb617874e70a/src/index.js#L35) isn't being reset because the parser is reused across webpack compilations. It would certainly explain the `0` --> `1` bug there.

Reply:
Been trying to debug further and have some findings. 

In our test suite we run webpack several times in the same thread. I think the test order affects the result of rebuilds with `worker-plugin`.

To test this I added a lot of debug logs and tried running the basic test and the rebuild test in different orders. Then I collected the logs just from the rebuild test.

One of the things I logged was the entry for the child compilation: https://github.com/GoogleChromeLabs/worker-plugin/blob/7ea61fb83a02170d0235e4590c4deb617874e70a/src/loader.js#L77. The others were build count, changed file being loaded, and which bundle contains the changed content.

If I run the rebuild test first and the basic test second I see:
- `dep.ts` is loaded with `bar`
- child compilation finishes for entry `0.worker.js`
- first build finishes
- `dep.ts` is loaded with `baz`
- child compilation finishes for entry `1.worker.js`
- second build finishes
- `0.worker.js` does not contains `baz`
- `1.worker.js` contains `baz`

If I run the rebuild test first and the basic test second I see:
- `dep.ts` is loaded with `bar`
- child compilation finishes for entry `0.worker.js`
- child compilation finishes for entry `1.worker.js`
- first build finishes
- `dep.ts` is loaded with `baz`
- child compilation finishes for entry `0.worker.js`
- second build finishes
- `0.worker.js` contains `baz`
- `1.worker.js` does not contains `baz`

So the end result is that the order of the tests seems to affect which bundle ends up containing the changed code.

But more interestingly, when the rebuild test is second, before the first build finishes there are a total of two child compilations that run. That part I find rather odd. I don't understand why that would happen. Even weirder is that the next child compilation is again the `0` one.